core/services/keeper/registry_synchronizer.go

```diff
@@ -0,0 +1,239 @@
+)
+
+const syncUpkeepQueueSize = 10
+
+func NewRegistrySynchronizer(
+	job job.Job,
+	contract *keeper_registry_wrapper.KeeperRegistry,
+	orm ORM,
+	syncInterval time.Duration,
+) *RegistrySynchronizer {
+	return &RegistrySynchronizer{
+		contract:      contract,
+		doneWG:        sync.WaitGroup{},
+		orm:           orm,
+		job:           job,
+		interval:      syncInterval,
+		chDone:        make(chan struct{}),
+		StartStopOnce: utils.StartStopOnce{},
+	}
+}
+
+// RegistrySynchronizer conforms to the job.Service interface
+var _ job.Service = &RegistrySynchronizer{}
+
+type RegistrySynchronizer struct {
+	contract *keeper_registry_wrapper.KeeperRegistry
+	doneWG   sync.WaitGroup
+	interval time.Duration
+	job      job.Job
+	orm      ORM
+
+	chDone chan struct{}
+
+	utils.StartStopOnce
+}
+
+func (rs *RegistrySynchronizer) Start() error {
+	if !rs.OkayToStart() {
+		return errors.New("RegistrySynchronizer is already started")
+	}
+	go rs.run()
+	return nil
+}
+
+func (rs *RegistrySynchronizer) Close() error {
+	if !rs.OkayToStop() {
+		return errors.New("RegistrySynchronizer is already stopped")
+	}
+	close(rs.chDone)
+	rs.doneWG.Wait()
+	return nil
+}
+
+func (rs *RegistrySynchronizer) run() {
+	rs.doneWG.Add(1)
+	ticker := time.NewTicker(rs.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-rs.chDone:
+			rs.doneWG.Done()
+			return
+		case <-ticker.C:
+			rs.syncRegistry()
+		}
+	}
+}
+
+func (rs *RegistrySynchronizer) syncRegistry() {
+	contractAddress := rs.job.KeeperSpec.ContractAddress
+	logger.Debugf("syncing registry %s", contractAddress.Hex())
+
+	err := func() error {
+		registry, err := rs.newSyncedRegistry(rs.job)
+		if err != nil {
+			return err
+		}
+		ctx, _ := postgres.DefaultQueryCtx()
+		if err = rs.orm.UpsertRegistry(ctx, &registry); err != nil {
+			return err
+		}
+		if err = rs.addNewUpkeeps(registry); err != nil {
+			return err
+		}
+		if err = rs.deleteCanceledUpkeeps(registry); err != nil {
+			return err
+		}
+		return nil
+	}()
+
+	if err != nil {
+		logger.Errorf("unable to sync registry %s, err: %v", contractAddress.Hex(), err)
+	}
+}
+
+func (rs *RegistrySynchronizer) addNewUpkeeps(reg Registry) error {
+	ctx, _ := postgres.DefaultQueryCtx()
+	nextUpkeepID, err := rs.orm.NextUpkeepIDForRegistry(ctx, reg)
+	if err != nil {
+		return errors.Wrap(err, "RegistrySynchronizer: unable to find next ID for registry")
+	}
+
+	countOnContractBig, err := rs.contract.GetUpkeepCount(nil)
+	if err != nil {
+		return errors.Wrapf(err, "RegistrySynchronizer: unable to get upkeep count")
+	}
+	countOnContract := countOnContractBig.Int64()
+
+	if nextUpkeepID > countOnContract {
+		return errors.New("RegistrySynchronizer: invariant, contract should always have at least as many upkeeps as DB")
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(int(countOnContract - nextUpkeepID))
+
+	// batch sync registries
+	chSyncUpkeepQueue := make(chan struct{}, syncUpkeepQueueSize)
+	done := func() { <-chSyncUpkeepQueue; wg.Done() }
```

Comment:
Is there a chance that this could block forever and stop the RegistrySynchronizer from shutting down? Maybe better to do something like:
```go
    ctx, cancel := utils.CombinedContext(ctx, rs.chDone)
    defer cancel()

    done := func() {
        select {
        case <-ctx.Done():
        case <-chSyncUpkeepQueue:
        }
        wg.Done()
    }
```

Reply:
also a good suggestion